### PR TITLE
Added Caption gen in 2 models

### LINKED ISSUE #665: Python Script to automatically add Captions to an image, using Deep learning
**Is your feature request related to a problem? Please describe.**
Deep learning is an interesting field, so many useful things are possible.
A script to automatically come up with a sensible caption is always good.

**Describe the solution you'd like**
The `alt text` is an important attribute of an image, and every image should have one.
Using Deep learning/Machine learning it is possible to actually come up with that type of text automatically.

**Describe alternatives you've considered**
There are a few online caption suggestors, but they are either paid (after few free attempts) or straight-up nonsensical

**Additional context**
Deep Learning is preferred as it is faster for these kinds of activities.

### PR DESCRIPTION
**Pull request by vybhav72954**
# Description

I have added a Caption Generator, along with the classification script, for additional help I have uploaded the Notebooks as well. This will ensure that people who want to dwell more in the working of the script can use the notebook and play around as well.

Moreover, all the models have been uploaded (Google Drive links). A model visualization along with details of the dataset and reasoning behind the usage of the models has been included.

FOr those who aren't really interested in working, but just want to use the script The README is there to help them.

Fixes #665 

Replace `issue_no` with the issue number which is fixed in this PR

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works

@HarshCasper Kindly deal with the labels in the issue as well, as this is actually a pretty old issue of mine.

### CODE REVIEW

**Comment by vybhav72954** [comment]
@HarshCasper I am not sure why there is a conflict here, I will try to handle it from my end.

**Comment by vybhav72954** [comment]
@HarshCasper I am not sure why there is a conflict here, I will try to handle it from my end.

UPDATE: AM really not sure whats going on here, can you see this?

**Comment by vybhav72954** [comment]
Ok its done and ready

**Comment by HarshCasper** [comment]
Can you please resolve the Merge Conflicts? 

In the meantime, I will review the changes.

**Comment by vybhav72954** [comment]
@HarshCasper The merge conflicts have been resolved, kindly see to this

File: Python/README.md [hunk 1]
```diff
@@ -173,3 +173,4 @@ S. No | Name of Script | Directory Link |  Author | Tools Used |
 163 | YouTube Video-Audio Download | [Click Here](/Python/YouTube_Video-Audio_Download) | [Naman Shah](https://github.com/namanshah01) | os, sys, pafy, requests
 164 | Zipper | [Click Here](/Python/Zipper) | [Puneet Pal Singh](https://github.com/Puneet-Pal-Singh) | os, zipfile, datetime
 165 | Zoom Automation | [Click Here](/Python/Zoom_Automation) | [Rutuja Dhanawade](https://github.com/rutujadhanawade) | pyautogui, time
+166 | Automatic Caption Generator for IMages | [Click Here](Python/Auto_Caption_Generator) | [Vybhav Chaturvedi](https://github.com/vybhav72954) | TensorFlow, Keras, inception, xception
```
Comment (HarshCasper) [inline Python/README.md:176]: You can revert these changes. We will bring up a new README

File: Python/Auto_Caption_Generator/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+oauthlib==3.1.0
+opt-einsum==3.3.0
+packaging==20.8
+pandas==1.2.0
+pandocfilters==1.4.3
+parso==0.8.1
+pickleshare==0.7.5
+Pillow==8.0.1
+pooch==1.3.0
+prometheus-client==0.9.0
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/requirements.txt:67]: Why are we using `prometheus-client` here :sweat_smile: 

Can you use this [PyPi Library](https://pypi.org/project/pip-check-reqs/) for the same?

File: Python/Auto_Caption_Generator/README.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+It is recommended to train these Neural Networks using GPU accelerated hardware.
+User first need to have a CUDA enabled Graphics Card, if this condition is met, Download CUDA toolkit and cuDNN library.
+
+For installation and help, these links are helpful:
+
+[LINK1](https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html)
+
+[LINK2](https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html)
+
+[LINK3](https://medium.com/swlh/cuda-installation-in-windows-2020-638b008b4639)
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/README.md:56]: Instead of `LINKN` can we use the actual names of the article? Would be better for understanding.

File: Python/Auto_Caption_Generator/caption_generator_Inception.py [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+
+# load doc into memory
+def load_doc(filename):
+    """
+    Function for loading various files
+    :param filename: Source Path
+    :return: Caption Text
+    """
+    # open the file as read only
+    file = open(filename, 'r')
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Inception.py:49]: You can use the `with` statement to open the file

File: Python/Auto_Caption_Generator/caption_generator_Inception.py [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+    Save descriptions in a file
+    :param descriptions: Clean descriptions
+    :param filename: Path of File
+    """
+    lines = list()
+    for key, desc_list in descriptions.items():
+        for desc in desc_list:
+            lines.append(key + ' ' + desc)
+    data = '\n'.join(lines)
+    file = open(filename, 'w')
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Inception.py:164]: You can use the `with` statement to open the file

File: Python/Auto_Caption_Generator/caption_generator_Xception.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    """
+
+
+    def load_doc(filename):
+        """
+        Function for loading various files
+        :param filename: Source Path
+        """
+        # Opening the file as read only
+        file = open(filename, 'r')
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Xception.py:47]: You can use the `with` statement to open the file

File: Python/Auto_Caption_Generator/caption_generator_Xception.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+        Save captions in one file
+        :param descriptions: Captions
+        :param filename: Destination file
+        """
+        lines = list()
+        for key, desc_list in descriptions.items():
+            for desc in desc_list:
+                lines.append(key + '\t' + desc)
+        data = "\n".join(lines)
+        file = open(filename, "w")
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Xception.py:122]: You can use the `with` statement to open the file

File: Python/Auto_Caption_Generator/caption_generator_Xception.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
+# -*- coding: utf-8 -*-
+
+import string
+import numpy as np
+from PIL import Image
+import os
+from pickle import dump, load
+import numpy as np
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Xception.py:8]: numpy has been imported twice

File: Python/Auto_Caption_Generator/caption_generator_Xception.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+    def text_vocabulary(descriptions):
+        """
+        Generate all the unique words from captions
+        :param descriptions: Captions
+        :return: Set of unique words
+        """
+        # build vocabulary of all unique words
+        vocab = set()
+        for key in descriptions.keys():
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Xception.py:106]: Consider using `descriptions.items()` to iterate over the dictionary

File: Python/Auto_Caption_Generator/caption_generator_Xception.py [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+
+
+    def dict_to_list(descriptions):
+        """
+        Convert Dictionary of captions to list
+        :param descriptions: Dict of captions
+        :return: List of captions
+        """
+        all_desc = []
+        for key in descriptions.keys():
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Xception.py:242]: Consider using `descriptions.items()` to iterate over the dictionary

File: Python/Auto_Caption_Generator/caption_generator_Inception.py [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+# convert the loaded descriptions into a vocabulary of words
+def to_vocabulary(descriptions):
+    """
+    Convert Captions into meaningful words
+    :param descriptions: Clean Description files
+    :return: Set of words corresponding to caption
+    """
+    # build a list of all description strings
+    all_desc = set()
+    for key in descriptions.keys():
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Inception.py:144]: Consider using `descriptions.items()` to iterate over the dictionary

File: Python/Auto_Caption_Generator/caption_generator_Inception.py [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+
+
+def to_lines(descriptions):
+    """
+    Convert Dictionary to List
+    :param descriptions: Clean Description
+    :return: List of all descriptions
+    """
+    all_desc = list()
+    for key in descriptions.keys():
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Inception.py:385]: Consider using `descriptions.items()` to iterate over the dictionary

File: Python/Auto_Caption_Generator/caption_generator_Inception.py [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+
+
+# parse descriptions
+descriptions = load_descriptions(doc)
+print('Loaded: %d ' % len(descriptions))
+
+
+def clean_descriptions(descriptions):
+    """
+    FUnction to Remove Extra  whitespaces, punctuation marks etc
```
Comment (HarshCasper) [inline Python/Auto_Caption_Generator/caption_generator_Inception.py:102]: Replace `FUnction` to `Function`

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by vybhav72954** [comment]
I deleted this branch already, I will close this PR and open a new one instead.

Let me go through all the suggestions then I will be in a position to comment on the possibilities.

File: Python/Auto_Caption_Generator/requirements.txt (hunk 2 above)
Comment (vybhav72954) [reply Python/Auto_Caption_Generator/requirements.txt:67]: That is one interesting Lib, I removed more than half of useless dependencies. Thanks!!!

File: Python/README.md (hunk 1 above)
Comment (vybhav72954) [reply Python/README.md:176]: Understood

File: Python/Auto_Caption_Generator/caption_generator_Inception.py (hunk 13 above)
Comment (vybhav72954) [reply Python/Auto_Caption_Generator/caption_generator_Inception.py:102]: While I was at this, I removed a few more spelling mistakes as well.

File: Python/Auto_Caption_Generator/caption_generator_Xception.py (hunk 8 above)
Comment (vybhav72954) [reply Python/Auto_Caption_Generator/caption_generator_Xception.py:8]: Thanks for pointing out

File: Python/Auto_Caption_Generator/caption_generator_Xception.py (hunk 9 above)
Comment (vybhav72954) [reply Python/Auto_Caption_Generator/caption_generator_Xception.py:106]: I don't understand the need for this one though, although it is a bit faster, I think the `keys` method is doing the best job here, I will make the changes though.

File: Python/Auto_Caption_Generator/README.md (hunk 3 above)
Comment (vybhav72954) [reply Python/Auto_Caption_Generator/README.md:56]: Done

**Comment by vybhav72954** [comment]
@HarshCasper I have migrated the PR to #798, do close this PR as soon as you accept that one.
